### PR TITLE
Conversion to Typescript

### PR DESCRIPTION
**Pull request by anmolarora1**
Addresses #519 

As a part of this PR, we'll introduce TS to util functions, add frequently used types from business-context to `src/types.ts` and other types shared by util functions to `src/utilTypes.ts`

### CODE REVIEW

File: src/utilTypes.ts [hunk 1]
```diff
@@ -3,3 +3,6 @@ export type ComparatorValue = 1 | -1 | 0
 
 /* A standard comparator function used within sort */
 export type ComparatorFunction<T> = (a: NonNullable<T>, b: NonNullable<T>) => ComparatorValue
+
+/* Set of file types supported for exporting thoughts */
+export type FileTypes = 'text/plain' | 'text/html'
```
Comment (raineorshine) [inline src/utilTypes.ts:8]: `MimeType` would be a more appropriate name.

A type should generally be in the singular. The plural would likely be its own array or collection type.

File: src/util/equalArrays.ts [hunk 2]
```diff
@@ -1,7 +1,5 @@
-//@ts-nocheck
-
 /** Equality for lists of lists. */
-export const equalArrays = (a, b) =>
+export const equalArrays = (a: unknown[], b: unknown[]): boolean =>
```
Comment (raineorshine) [inline src/util/equalArrays.ts:2]: Does `a: []` work, or is `unknown` required?

File: src/util/equalPath.ts [hunk 3]
```diff
@@ -1,7 +1,6 @@
-//@ts-nocheck
-
 import { equalThoughtRanked } from './equalThoughtRanked'
+import { Thought, Child } from '../types'
 
 /** Compares two thoughtsRanked arrays using { value, rank } as identity and ignoring other properties. */
-export const equalPath = (a, b) =>
+export const equalPath = (a: (Thought | Child)[], b: (Thought | Child)[]): boolean =>
```
Comment (raineorshine) [inline src/util/equalPath.ts:5]: `a` and `b` should be type `Path`.

File: src/util/equalThoughtRanked.ts [hunk 4]
```diff
@@ -1,5 +1,5 @@
-//@ts-nocheck
+import { Thought, Child } from "../types";
 
 /** Compares two thought objects using { value, rank } as identity and ignoring other properties. */
-export const equalThoughtRanked = (a, b) =>
+export const equalThoughtRanked = (a: (Thought | Child), b: (Thought | Child)): boolean =>
```
Comment (raineorshine) [inline src/util/equalThoughtRanked.ts:4]: `Child` only

File: src/util/equalThoughtSorted.ts [hunk 5]
```diff
@@ -1,5 +1,5 @@
-//@ts-nocheck
+import { Thought, Child } from "../types";
 
 /** Compares two thought objects using { value } as identity and ignoring other properties. */
-export const equalThoughtSorted = (a, b) =>
+export const equalThoughtSorted = (a: (Thought | Child), b: (Thought | Child)): boolean =>
```
Comment (raineorshine) [inline src/util/equalThoughtSorted.ts:4]: `Child` only

File: src/util/equalArrays.ts (hunk 2 above)
Comment (anmolarora1) [reply src/util/equalArrays.ts:2]: Afaik, that doesn't work. We need to specify the type of items.

File: src/utilTypes.ts (hunk 1 above)
Comment (anmolarora1) [reply src/utilTypes.ts:8]: Agreed

File: src/util/formatNumber.ts [hunk 6]
```diff
@@ -1,11 +1,9 @@
-//@ts-nocheck
-
 /** Adds commas to a number */
 // TODO: Localize
-export const formatNumber = n => {
+export const formatNumber = (n: number) => {
   let s = '' // eslint-disable-line fp/no-let
   const digits = n.toString().split('')
-  digits.forEach((value, i) => {
+  digits.forEach((_, i) => {
```
Comment (anmolarora1) [inline src/util/formatNumber.ts:6]: @raineorshine Have you considered using `toLocaleString` instead of this?

File: src/util/equalArrays.ts (hunk 2 above)
Comment (raineorshine) [reply src/util/equalArrays.ts:2]: Did you try it?

File: src/util/formatNumber.ts (hunk 6 above)
Comment (raineorshine) [reply src/util/formatNumber.ts]: Sure, I'm not super familiar with `toLocaleString`.

Comment (anmolarora1) [reply src/util/formatNumber.ts]: It does the same job that this function is doing. You can check it out [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toLocaleString)

File: src/util/equalArrays.ts (hunk 2 above)
Comment (anmolarora1) [reply src/util/equalArrays.ts:2]: Yes. Please check [this](https://www.typescriptlang.org/play/#code/MYewdgzgLgBA5gUygGQWOUAWMC8MAUAhgFwwDaAugJS4B8MhAdADZoaYDcAUF6JCKxYg4+RCjZZ8ZAIwAaAEzUqQA)

Comment (raineorshine) [reply src/util/equalArrays.ts:2]: Got it, thanks!

File: src/util/formatNumber.ts (hunk 6 above)
Comment (raineorshine) [reply src/util/formatNumber.ts]: Cool! I created #682 to replace it.

File: package-lock.json [hunk 7]
```diff
@@ -8623,25 +8623,25 @@
           "dependencies": {
             "abbrev": {
               "version": "1.1.1",
-              "resolved": "https://registry.npmjs.org/abbrev/-/abbrev-1.1.1.tgz",
+              "resolved": false,
```
Comment (raineorshine) [inline package-lock.json:8626]: Revert `package-lock.json`

File: src/util/contextOf.ts [hunk 8]
```diff
@@ -1,4 +1,5 @@
-import { Context } from "../types";
+import { Thought, Child } from "../types";
 
+// type C = 
 /** Gets the context of a context */
-export const contextOf = (context: Context): Context => context.slice(0, -1)
+export const contextOf = <T extends Array<Thought | Child>>(context: T) => context.slice(0, -1)
```
Comment (raineorshine) [inline src/util/contextOf.ts]: I think this might need to be `Context | Path`.

File: src/util/deleteThought.ts [hunk 9]
```diff
@@ -37,17 +35,18 @@ import {
...
   splitChain,
 } from '../selectors'
+import { Child, Thought } from '../types'
 
 /** Deletes a thought. */
 export const deleteThought = () => {
 
   const state = store.getState()
-  const path = state.cursor
+  const path:(Thought[] | Child[]) = state.cursor
```
Comment (raineorshine) [inline src/util/deleteThought.ts:44]: Why use `Child[]` when we have a `Path` type? Also, why `Thought[]` at all? A `Thought` is an entry in the `thoughtIndex` which is not related to a `Path`.

File: src/util/deleteThought.ts [hunk 10]
```diff
@@ -37,17 +35,18 @@ import {
...
 
   // same as in newThought
   const contextChain = splitChain(state, path)
   const showContexts = isContextViewActive(state, contextOf(path))
-  const thoughtsRanked = contextChain.length > 1
+  const thoughtsRanked:Thought[] = contextChain.length > 1
```
Comment (raineorshine) [inline src/util/deleteThought.ts:49]: Doesn't Typescript infer types of variables?

File: src/util/deleteThought.ts [hunk 11]
```diff
@@ -73,12 +72,12 @@ export const deleteThought = () => {
...
 
   // prev must be calculated before dispatching existingThoughtDelete
-  const prev = showContexts
+  const prev: Child | undefined = showContexts
     ? prevContext()
     : prevSibling(state, value, context, rank)
 
   /** Returns true when thought is not hidden due to being a function or having a =hidden attribute. */
-  const isVisible = thoughtRanked => state.showHiddenThoughts || (
+  const isVisible = (thoughtRanked: Thought) => state.showHiddenThoughts || (
```
Comment (raineorshine) [inline src/util/deleteThought.ts:80]: I think this should be `Child`.

File: src/util/deleteThought.ts [hunk 12]
```diff
@@ -103,7 +102,7 @@ export const deleteThought = () => {
   }
 
   /** Sets the cursor or moves it back if it doesn't exist. */
-  const setCursorOrBack = (thoughtsRanked, { offset } = {}) => {
+  const setCursorOrBack = (thoughtsRanked: Array<Thought | Child | string> = [], { offset } = {offset: 0}) => {
```
Comment (raineorshine) [inline src/util/deleteThought.ts:105]: I think there are times when `offset` is undefined. If you set it to `0` it will change the behavior.

Comment (raineorshine) [inline src/util/deleteThought.ts:105]: Wouldn't this type just be a nullable `Path`?

File: src/util/deleteThought.ts [hunk 13]
```diff
@@ -112,14 +111,16 @@ export const deleteThought = () => {
...
     // Case III: delete last thought in context; set cursor on context
-    thoughts.length > 1 ? [rootedContextOf(path), { offset: head(context).length }]
-    // Case IV: delete very last thought; remove cursor
-    : [null]
-  )
+    : setCursorOrBack(contextOf(path).concat(next()), { offset: 0 })
+  }
+  // Case IV: delete very last thought; remove cursor
+  return thoughts.length>1 ?setCursorOrBack(rootedContextOf(path), { offset: head(context).length }) : setCursorOrBack([])
 }
```
Comment (raineorshine) [inline src/util/deleteThought.ts:126]: Do not use multiple return statements, as it breaks the invariant that the shape of the code should reflect the control flow. A single return statement with multiple cases captures the logic with less redundancy.

File: src/selectors/canShowModal.ts [hunk 14]
```diff
@@ -1,6 +1,6 @@
 import { store } from '../store'
 
-const canShowModal = (state: any, id: number) => {
+const canShowModal = (state: any, id: number | string) => {
```
Comment (raineorshine) [inline src/selectors/canShowModal.ts:3]: Would this work with just `string`?

File: src/util/ancestors.ts [hunk 15]
```diff
@@ -0,0 +1,6 @@
+import { equalThoughtRanked } from './equalThoughtRanked'
+import { Path,Child } from '../types'
+
+
```
Comment (raineorshine) [inline src/util/ancestors.ts:4]: Extra blank line. Can you make sure the typescript linter is picking this up?

File: src/util/ancestors.ts [hunk 16]
```diff
@@ -0,0 +1,6 @@
+import { equalThoughtRanked } from './equalThoughtRanked'
+import { Path,Child } from '../types'
```
Comment (raineorshine) [inline src/util/ancestors.ts:2]: Space after comma

File: src/util/compareThought.ts [hunk 17]
```diff
@@ -104,4 +106,4 @@ const compareReasonable = makeOrderedComparator([
 ])
 
 /** Compare the value of two thoughts. */
-export const compareThought = (a, b) => compareReasonable(removeEmojis(a.value), removeEmojis(b.value))
+export const compareThought = (a: Thought, b: Thought) => compareReasonable(removeEmojis(a.value), removeEmojis(b.value))
```
Comment (raineorshine) [inline src/util/compareThought.ts:109]: `Child`

File: src/util/decodeCharacterEntities.ts [hunk 18]
```diff
@@ -1,5 +1,5 @@
 /** Decode &, >, and < character entities. */
-export const decodeCharacterEntities = s => s
+export const decodeCharacterEntities = (s:string): string => s
```
Comment (raineorshine) [inline src/util/decodeCharacterEntities.ts:2]: Space after `:`. Can the linter catch this? (Can be in another PR)

File: src/util/contextOf.ts [hunk 19]
```diff
@@ -0,0 +1,10 @@
+import { Context, Path } from "../types";
+
+type ReturnType = {
+  (param:Context): Context;
+  (param:Path): Path;
+  (param: Path | Context) : Path | Context
+}
```
Comment (raineorshine) [inline src/util/contextOf.ts:7]: Can you explain this? I don't quite understand.

File: src/util/ellipsize.ts [hunk 20]
```diff
@@ -1,6 +1,6 @@
 import { THOUGHT_ELLIPSIZED_CHARS } from '../constants'
 
 /** Returns a string truncated with an ellipsis at a given limit n. */
-export const ellipsize = (s, n = THOUGHT_ELLIPSIZED_CHARS) =>
+export const ellipsize = (s: string, n:number = THOUGHT_ELLIPSIZED_CHARS): string =>
```
Comment (raineorshine) [inline src/util/ellipsize.ts:4]: Space after `:`

File: src/util/ellipsizeUrl.ts [hunk 21]
```diff
@@ -2,7 +2,7 @@ import { URL_MAX_CHARS } from '../constants'
 import { isURL } from '../util/isURL'
 
 /** Ellipsize a value if it is a url. */
-export const ellipsizeUrl = value =>
+export const ellipsizeUrl = (value:string): string =>
```
Comment (raineorshine) [inline src/util/ellipsizeUrl.ts:5]: Space after `:`

File: src/util/escapeRegExp.ts [hunk 22]
```diff
@@ -1,4 +1,4 @@
 /** Make a string safe to instantiate a RegExp.
  * NOTE: The escape-string-regexp npm package crashes the react production build.
  */
-export const escapeRegExp = s => s.replace(/[-[\]{}()*+?.\\^$|#\s]/g, '\\$&')
+export const escapeRegExp = (s:string): string => s.replace(/[-[\]{}()*+?.\\^$|#\s]/g, '\\$&')
```
Comment (raineorshine) [inline src/util/escapeRegExp.ts:4]: Space after `:`

File: src/util/escapeSelector.ts [hunk 23]
```diff
@@ -3,4 +3,4 @@ import { escapeRegExp } from './escapeRegExp'
 export const regExpEscapeSelector = new RegExp('[' + escapeRegExp(' !"#$%&\'()*+,./:;<=>?@[]^`{|}~') + ']', 'g')
 
 /** Replace characters that are invalid in document.querySelector with their respective character codes. Prepend _ to escape leading digits. */
-export const escapeSelector = s => '_' + s.replace(regExpEscapeSelector, s => '_' + s.charCodeAt())
+export const escapeSelector = (s:string): string => '_' + s.replace(regExpEscapeSelector, s => `_${s.charCodeAt(0)}`)
```
Comment (raineorshine) [inline src/util/escapeSelector.ts:6]: Space after `:`

File: src/util/formatNumber.ts [hunk 24]
```diff
@@ -1,3 +1,3 @@
 /** Adds commas to a number. */
 // TODO: Localize
-export const formatNumber = (n: number, locale='en-US') => n.toLocaleString(locale)
+export const formatNumber = (n: number, locale='en-US') => n.toLocaleString(locale)
```
Comment (raineorshine) [inline src/util/formatNumber.ts:3]: Restore newline at end of file.

Do you have `EditorConfig` configured in your editor? Usually that will automatically enforce this.

File: src/util/makeCompareByProp.ts [hunk 25]
```diff
@@ -1,16 +1,18 @@
...
 /**
  * Creates a function that takes two values and compares the given key.
  * Does case insensitive comparison with strings.
  */
-export const makeCompareByProp = key => (a, b) =>
+export const makeCompareByProp = (key:any):ComparatorFunction<Object> => (a: Object, b: Object) =>
   isGreater(a, b, key) ? 1
   : isSmaller(a, b, key) ? -1
-  : 0
+  : 0
```
Comment (raineorshine) [inline src/util/makeCompareByProp.ts:18]: Newline

File: src/util/makeCompareByProp.ts [hunk 26]
```diff
@@ -1,16 +1,18 @@
...
 
+import { ComparatorFunction } from '../utilTypes'
+
 /** Returns true if the first object's key is greater than the second object's key. */
-export const isGreater = (a, b, key) => lower(a[key]) > lower(b[key])
+export const isGreater = (a:Object, b:Object, key: keyof typeof a & keyof typeof b )  => lower(a[key]) > lower(b[key])
 
 /** Returns true if the first object's key is smaller than the second object's key. */
-export const isSmaller = (a, b, key) => lower(a[key]) < lower(b[key])
+export const isSmaller = (a:Object, b:Object, key: keyof typeof a & keyof typeof b )  => lower(a[key]) < lower(b[key])
```
Comment (raineorshine) [inline src/util/makeCompareByProp.ts:9]: Space after `:`

File: src/util/makeCompareByProp.ts [hunk 27]
```diff
@@ -1,16 +1,18 @@
 import { lower } from './lower'
 
+import { ComparatorFunction } from '../utilTypes'
+
 /** Returns true if the first object's key is greater than the second object's key. */
-export const isGreater = (a, b, key) => lower(a[key]) > lower(b[key])
+export const isGreater = (a:Object, b:Object, key: keyof typeof a & keyof typeof b )  => lower(a[key]) > lower(b[key])
```
Comment (raineorshine) [inline src/util/makeCompareByProp.ts:6]: Can you explain the expression `keyof typeof a & keyof typeof b`?

File: src/utilTypes.ts [hunk 28]
```diff
@@ -0,0 +1,13 @@
...
+/* A standard comparator function used within sort */
+export type ComparatorFunction<T> = (a: NonNullable<T>, b: NonNullable<T>) => ComparatorValue
+
+/* Set of file types supported for exporting thoughts */
+export type MimeType = 'text/plain' | 'text/html'
+
+/* Generic type to allow null */
+export type Nullable<T> = T | null;
+
+export type GenericObject<T = any> = {[key: string]: T}
```
Comment (raineorshine) [inline src/utilTypes.ts:13]: What's the advantage of using this over any? Since everything is a generic object in Javascript.

File: src/util/parseJsonSafe.ts [hunk 29]
```diff
@@ -1,8 +1,10 @@
+import { Nullable } from "../utilTypes"
+
 /** Safely JSON parse an unknown value, and default to a given fallback value. */
-export const parseJsonSafe = (value, defaultValue) => {
+export const parseJsonSafe = (value: Nullable<string>, defaultValue: number) => {
```
Comment (raineorshine) [inline src/util/parseJsonSafe.ts:4]: `defaultValue` should be `any`.

File: src/util/pathToContext.ts [hunk 30]
```diff
@@ -0,0 +1,8 @@
+import { Context, Path } from "../types"
+
+/** Converts paths [{ value, rank }, ...] to contexts [key, ...]. */
+// if already converted, return a shallow copy
+// if falsey, return as-is
+
+const isPath = (param: any): param is Path =>  !!(param as Path).length && (param as Path)[0].hasOwnProperty('value')
+export const pathToContext = (thoughts: Path | Context): Context  => isPath(thoughts) ? thoughts.map(child => child.value) : thoughts.slice()
```
Comment (raineorshine) [inline src/util/pathToContext.ts:8]: Newline

File: src/util/rankThoughtsSequential.ts [hunk 31]
```diff
@@ -1,3 +1,5 @@
+import { Context, Child } from "../types";
+
 /** Ranks the thoughts from 0 to n in the given array order. */
-export const rankThoughtsSequential = thoughts =>
+export const rankThoughtsSequential = (thoughts: Context) =>
```
Comment (raineorshine) [inline src/util/rankThoughtsSequential.ts:4]: Technically this should be `string[]` since this operates on thoughts within a context rather than along a path.

File: src/util/setSelection.ts [hunk 32]
```diff
@@ -5,20 +5,23 @@
...
-  const sel = window.getSelection()
+  const sel = window.getSelection() || new Selection()
   // automatically constrain offset to text length
   // this may still throw an error if the text node does no exist any longer
-  try {
-    range.setStart(textNode, offset ? Math.min(offset, textNode.textContent.length) : end ? textNode.textContent.length : 0)
-  }
-  catch (e) {
-    console.warn(e)
+  if(textNode!==null) {
```
Comment (raineorshine) [inline src/util/setSelection.ts:18]: I believe the `textNode` is guaranteed to be valid since it is appended at the beginning of the function.

File: src/util/setSelection.ts [hunk 33]
```diff
@@ -5,20 +5,23 @@
...
-export const setSelection = (el, { offset, end } = {}) => {
+type SelectionOptionsType = {offset: number, end: number}
+export const setSelection = (el: HTMLElement, { offset, end }: SelectionOptionsType = {offset: 0, end: 0}) => {
   if (el.childNodes.length === 0) {
     el.appendChild(document.createTextNode(''))
   }
   const textNode = el.firstChild
   const range = document.createRange()
-  const sel = window.getSelection()
+  const sel = window.getSelection() || new Selection()
```
Comment (raineorshine) [inline src/util/setSelection.ts:15]: Can you explain this?

**Comment by raineorshine** [comment]
> - isRoot & isEm are being used interchangeably by contexts/Path/Thought object’s context field. What’s the correct expectation?

`isRoot` tests if a thought is a descendant of the root thought. `isEm` tests if a thought is a descendant of the **em** thought. Currently the root thought contains user thoughts and **em** contains Settings.

Does that answer your question, or are you seeking something else?

> - mergeUpdates - why are we using for in and not Object.keys?

For [performance reasons](https://jsperf.com/object-keys-vs-for-in-loop-test), but this will become less important when we switch to a thought cache (#273) rather than loading all thoughts into memory.

> - notNull function doesn’t seem to be filtering undefined whereas its JSDoc description states it filters both null and undefined.

`a != null` is shorthand for `a !== null && a !== undefined`

> - importHTML - line 92 . Is null a valid value in thoughtindexupdates?

The line number has changed, but I think I can still answer your question. Yes, `thoughtIndexUpdates` may have `null` values. This is what tells Firebase to delete the key. They are removed locally into `thoughtIndex` via `mergeUpdates`. Same for `contextIndexUpdates`.

> - What's the type of items in sumSubthoughtsLength?

That's a strange one. `Child | ThoughtContext`.

File: src/selectors/canShowModal.ts (hunk 14 above)
Comment (anmolarora1) [reply src/selectors/canShowModal.ts:3]: Yes, I'll update it

File: src/util/ancestors.ts (hunk 15 above)
Comment (anmolarora1) [reply src/util/ancestors.ts:4]: We can add a config for picking up such issues. If it's a long process, we can tackle it in a different issue.

File: src/util/contextOf.ts (hunk 19 above)
Comment (anmolarora1) [reply src/util/contextOf.ts:7]: An explicit return type is added here to allow the compiler to infer the type of results in the calling functions. If we don't do it, the return type would be `Path | Context`.  
An example from `importHtml.ts` - 
[importCursor](https://github.com/cybersemics/em/blob/adab4290f2a0c59d6fc53f52a3dd8040b3f21dba/src/util/importHTML.js#L98) can be `Path|Context`. But [here](https://github.com/cybersemics/em/blob/adab4290f2a0c59d6fc53f52a3dd8040b3f21dba/src/util/importHTML.js#L182), we're assuming its type to be `Path`, which results in a compile-time error.

File: src/utilTypes.ts (hunk 28 above)
Comment (anmolarora1) [reply src/utilTypes.ts:13]: That's true, everything is a generic object in JS and we can allow them to be like that. But this helps to narrow down the allowed value types for an object. 
Example - We can enforce the thoughtIndex to have only `thoughtObject` values - [`thoughtIndex: GenericObject<Thought>`](https://github.com/cybersemics/em/blob/5226dcfec237353436ae84973c23cacfa2b76764/src/util/initialState.ts#L33).

File: src/util/setSelection.ts (hunk 32 above)
Comment (anmolarora1) [reply src/util/setSelection.ts:18]: That's right, but the compiler doesn't quite understand that. 
Maybe we can refactor our code to make it more explicit. Something like this helps to get rid of the compiler error - 
```
const textNode = el.firstChild || el.appendChild(document.createTextNode(''))
```

File: src/util/setSelection.ts (hunk 33 above)
Comment (anmolarora1) [reply src/util/setSelection.ts:15]: When used in Firefox, `window.getSelection()` [returns null](https://developer.mozilla.org/en-US/docs/Web/API/Window/getSelection) for some values, and hence, TS compilation errors on [these lines](https://github.com/cybersemics/em/blob/094828d323f63e1d19d865947e619322c9dc2519/src/util/setSelection.ts#L27)

File: src/util/makeCompareByProp.ts (hunk 27 above)
Comment (anmolarora1) [reply src/util/makeCompareByProp.ts]: The intention behind adding this was to ensure that the key passed to this function belongs to both the objects, so it doesn't end up doing `lower(undefined) >lower(undefined)`
It didn't work as expected but it's fixed in [this](https://github.com/cybersemics/em/pull/656/commits/fb102a1254eee7ce8d99aa6157aa3606b16ce29b) commit
